### PR TITLE
Implement `GT.pipe()` and `GT.pipes()`

### PR DESCRIPTION
**Pull request by jrycw**
Related to #353.

After reviewing the `pipe()` methods from [Pandas](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.pipe.html) and [Polars](https://docs.pola.rs/py-polars/html/reference/dataframe/api/polars.DataFrame.pipe.html), I propose adding `GT.pipe()` and `GT.pipes()` in this PR.

### `GT.pipe()`
`GT.pipe()` will function similarly to `Pandas` and `Polars`, receiving a function along with optional positional and keyword arguments. An example is provided below:

```python
import polars as pl
from great_tables import GT, loc, style
from great_tables.data import towny


towny_mini = pl.from_pandas(towny).head(10)
columns = ["land_area_km2", "density_2021"]
colors = ["lightgray", "lightblue"]


def tbl_style(gtbl: GT, columns: list[str], colors: list[str]) -> GT:
    for column, color in zip(columns, colors):
        gtbl = gtbl.tab_style(
            style=style.fill(color=color),
            locations=loc.body(columns=column, rows=pl.col(column).eq(pl.col(column).max())),
        )
    return gtbl


(
    GT(
        towny_mini[["name", "land_area_km2", "density_2021"]],
        rowname_col="name",
    ).pipe(tbl_style, columns, colors)
)
```

### `GT.pipes()`
`GT.pipe()` serves as a convenient helper method for chaining multiple pre-built functions (or using `partial`). An example is provided below:

```python
from functools import partial
import polars as pl
from great_tables import GT, loc, style
from great_tables.data import towny


towny_mini = pl.from_pandas(towny).head(10)
columns = ["land_area_km2", "density_2021"]
colors = ["lightgray", "lightblue"]


def tbl_style(gtbl: GT, column: str, color: str) -> GT:
    return gtbl.tab_style(
        style=style.fill(color=color),
        locations=loc.body(columns=column, rows=pl.col(column).eq(pl.col(column).max())),
    )


(
    GT(
        towny_mini[["name", "land_area_km2", "density_2021"]],
        rowname_col="name",
    ).pipes(
        *[partial(tbl_style, column=column, color=color) for column, color in zip(columns, colors)]
    )
)
```
### Table
Both methods yield the table as follows:
image

### Assistance Required
I hope the API design makes sense, and could the team please review my docs to ensure they render correctly? I'm relatively new to `quarto`.

### CODE REVIEW

**Comment by jrycw** [comment]
I'm not sure if it's a good idea to support passing a list into `pipes()`. If we add two lines to `pipe()`, like:
```python
def pipes(self: "GT", *funcs: Callable["GT", "GT"] | list[Callable["GT", "GT"]]) -> "GT":
    if isinstance(funcs, list) and len(funcs) == 1:
        funcs = funcs[0]
    for func in funcs:
        self = pipe(self, func)
    return self
```

Then, we can do:

```python
(
    GT(
        towny_mini[["name", "land_area_km2", "density_2021"]],
        rowname_col="name",
    ).pipes(
        [partial(tbl_style, column=column, color=color) for column, color in zip(columns, colors)]
    )
)
```
@machow and @rich-iannone , could you share your thoughts on this idea with me?

**Comment by jrycw** [comment]
I've updated `.pipes()` to accept a list of functions in the latest commit. We can decide which commit to use later.

**Review by machow — CHANGES_REQUESTED** [review]
Sorry for the wait! Thanks for taking the time to work through this, and look at what pandas and polars do. I'm noticing pandas and polars have only the one `.pipe()` method. Maybe we start with this, and then if people are asking for it, consider a `.pipes()` method?

Packages like `toolz` have a function---called [`compose()`](https://toolz.readthedocs.io/en/latest/api.html#toolz.functoolz.compose)---that makes it easy to send multiple things to a single `.pipe()` method. For example...

```python
from toolz.functoolz import compose, compose_left

from functools import partial
import polars as pl
from great_tables import GT, loc, style
from great_tables.data import towny


towny_mini = pl.from_pandas(towny).head(10)
columns = ["land_area_km2", "density_2021"]
colors = ["lightgray", "lightblue"]


def tbl_style(gtbl: GT, column: str, color: str) -> GT:
    return gtbl.tab_style(
        style=style.fill(color=color),
        locations=loc.body(columns=column, rows=pl.col(column).eq(pl.col(column).max())),
    )


(
    GT(
        towny_mini[["name", "land_area_km2", "density_2021"]],
        rowname_col="name",
    ).pipe(
        compose_left(*[partial(tbl_style, column=column, color=color) for column, color in zip(columns, colors)])
    )
)
```

**Comment by jrycw** [comment]
@machow thanks for introducing me to `toolz`. It's new to me, and it looks like it has a lot of great features. Sure, let's remove the `.pipes()` method. I've reviewed the preview docs and made some adjustments. Other than that, it looks good to me. What do you think?

**Review by machow — CHANGES_REQUESTED** [review]
It's looking really good, I left a quick thought on how we might be able to annotate pipe using ParamSpec. But overall, it seems basically ready to go! @rich-iannone you'll probably know the right section for `.pipe` in the reference

File: great_tables/_pipe.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Callable, Any
+
+if TYPE_CHECKING:
+    from .gt import GT
+
+
+def pipe(self: "GT", func: Callable[..., "GT"], *args: Any, **kwargs: Any) -> "GT":
```
[the remark this answers is not in the document]
Comment (jrycw) [reply great_tables/_pipe.py]: Of course! This is somewhat new to me, so please feel free to correct me if I've misunderstood anything.

[the remark this answers is not in the document]
Comment (jrycw) [reply great_tables/_pipe.py]: @machow and @rich-iannone, this PR has been pending for a while. Could we schedule it for merging, or are there any remaining concerns?

File: docs/_quarto.yml [hunk 2]
```diff
@@ -174,6 +174,13 @@ quartodoc:
       contents:
         - GT.save
         - GT.as_raw_html
+    - title: Pipeline
```
[the remark this answers is not in the document]
Comment (rich-iannone) [reply docs/_quarto.yml:177]: I like it!

**Comment by machow** [comment]
Sorry for the wait 😓, Rich and I are pairing right now -- let me try to get this merged right now

**Review by machow — APPROVED** [review]
This LGTM, @rich-iannone do you want to merge? I simplified the test a bit, and think the lil pipeline section in the reference seems okay as is!

**Comment by rich-iannone** [comment]
Thank you @jrycw !

**Comment by jrycw** [comment]
> This LGTM, @rich-iannone do you want to merge? I simplified the test a bit, and think the lil pipeline section in the reference seems okay as is!

@machow, thanks for simplifying the test—it’s much more concise now.

**Comment by jrycw** [comment]
@machow  and @rich-iannone, I’ve removed `GT.pipes()` from the PR title since we didn’t end up merging that part into the codebase. This should help prevent potential confusion in the future.

**Pull request merged**